File: nand2tetris/projects/07/Rust/vmtranslator/src/encoder.rs
use crate::Arithmetic;
use crate::Command;
use crate::Segment;

pub struct Encoder {
    pub basename: String,
    pub counter: i32,
}

impl Encoder {
    pub fn exec(&mut self, cmd: Command) -> String {
        match cmd {
            Command::Arithmetic(x) => {
                let pre_binary_op =
                    [ "@SP"
                    , "AM=M-1"
                    , "D=M"
                    , "A=A-1"
                    ].join("\n");
                let pre_unary_op =
                    [ "@SP"
                    , "A=M-1"
                    ].join("\n");
                let mut comparison = |op: &str| {
                    let res =
                        [ "D=M-D"
                        , &*self.at_label("TRUE")
                        , &*["D;J", op].join("")
                        , "@SP"
                        , "A=M-1"
                        , "M=0"
                        , &*self.at_label("END")
                        , "0;JMP"
                        , &*self.label("TRUE")
                        , "@SP"
                        , "A=M-1"
                        , "M=-1"
                        , &*self.label("END")
                        ].join("\n");
                    self.counter += 1;
                    res
                };
                match x {
                    Arithmetic::Add => [&*pre_binary_op, "M=D+M"].join("\n"),
                    Arithmetic::Sub => [&*pre_binary_op, "M=M-D"].join("\n"),
                    Arithmetic::Neg => [&*pre_unary_op,  "M=-M"] .join("\n"),
                    Arithmetic::Eq  => [&*pre_binary_op, &*comparison("EQ")].join("\n"),
                    Arithmetic::Gt  => [&*pre_binary_op, &*comparison("GT")].join("\n"),
                    Arithmetic::Lt  => [&*pre_binary_op, &*comparison("LT")].join("\n"),
                    Arithmetic::And => [&*pre_binary_op, "M=D&M"].join("\n"),
                    Arithmetic::Or  => [&*pre_binary_op, "M=D|M"].join("\n"),
                    Arithmetic::Not => [&*pre_unary_op,  "M=!M"].join("\n"),
                }
            },
            Command::Push { segment, idx } => {
                let push_d =
                    [ "@SP"
                    , "A=M"
                    , "M=D"
                    , "@SP"
                    , "M=M+1"].join("\n");
                let push_seg = | seg: &str, idx: i32| {
                    [ &*at_const(idx)
                    , "D=A"
                    , &*["@", seg].join("")
                    , "A=D+M"
                    , "D=M"
                    , &*push_d
                    ].join("\n")
                };
                let push_seg_at = | at: i32, idx: i32| {
                    [ &*at_const(idx)
                    , "D=A"
                    , &*at_const(at)
                    , "A=D+A"
                    , "D=M"
                    , &*push_d
                    ].join("\n")
                };
                match segment {
                    Segment::Constant =>
                        [ &*at_const(idx)
                        , "D=A"
                        , &*push_d
                        ].join("\n"),
                    Segment::Local => push_seg("LCL", idx),
                    Segment::Argument => push_seg("ARG", idx),
                    Segment::This => push_seg("THIS", idx),
                    Segment::That => push_seg("THAT", idx),
                    Segment::Pointer => push_seg_at(3, idx),
                    Segment::Temp => push_seg_at(5, idx),
                    Segment::Static =>
                        [ &*["@", &*self.basename, ".", &*idx.to_string()].join("")
                        , "D=M"
                        , &*push_d
                        ].join("\n"),
                }
            },
            Command::Pop { segment, idx } => {
                let pop_d =
                    [ "@SP"
                    , "AM=M-1"
                    , "D=M"].join("\n");
                let d_to_seg =
                    [ "@R13"
                    , "A=M"
                    , "M=D"].join("\n");
                let pop_seg = | seg: &str, idx: i32| {
                    [ &*at_const(idx)
                    , "D=A"
                    , &*["@", seg].join("")
                    , "D=D+M"
                    , "@R13"
                    , "M=D"
                    , &*pop_d
                    , &*d_to_seg
                    ].join("\n")
                };
                let pop_seg_at = | at: i32, idx: i32| {
                    [ &*at_const(idx)
                    , "D=A"
                    , &*at_const(at)
                    , "D=D+A"
                    , "@R13"
                    , "M=D"
                    , &*pop_d
                    , &*d_to_seg
                    ].join("\n")
                };
                match segment {
                    Segment::Constant => pop_d,
                    Segment::Local => pop_seg("LCL", idx),
                    Segment::Argument => pop_seg("ARG", idx),
                    Segment::This => pop_seg("THIS", idx),
                    Segment::That => pop_seg("THAT", idx),
                    Segment::Pointer => pop_seg_at(3, idx),
                    Segment::Temp => pop_seg_at(5, idx),
                    Segment::Static =>
                        [ &*pop_d
                        , &*["@", &*self.basename, ".", &*idx.to_string()].join("")
                        , "M=D"
                        ].join("\n"),
                }
            },
            _ => unreachable!(),
            // Command::Label(symbol) => ,
            // Command::Goto(symbol) => ,
            // Command::If(symbol) => ,
            // Command::Function { symbol, num } => ,
            // Command::Call { symbol, num } => ,
            // Command::Return => ,
        }
    }
    fn symbol(&self, s: &str) -> String {
        [s, &*self.basename, &*self.counter.to_string()].join(".")
    }
    fn at_label(&self, s: &str) -> String {
        ["@", &*self.symbol(s)].join("")
    }
    fn label(&self, s: &str) -> String {
        ["(", &*self.symbol(s), ")"].join("")
    }
}

fn at_const(addr: i32) -> String {
    ["@", &*addr.to_string()].join("")
}

File: projects/07/Haskell/vmtranslator/src/Encoder.hs
{-# LANGUAGE OverloadedStrings #-}

module Encoder
    ( encode
    , BaseName
    ) where

import Numeric                          ( showInt )
import Data.Bifunctor                   ( bimap )
import qualified Data.Text as T

import Lib.Command
-- import Lib.Symbol
import Lib.Addr
import Lib.Arithmetic
import Lib.Segment

type BaseName = String

encode :: BaseName -> [Command] -> [T.Text]
encode f cs = encode' f 0 cs []

encode' :: BaseName -> Addr -> [Command] -> [T.Text] -> [T.Text]
encode' _ _ [] out     = reverse out
encode' f j (c:cs) out = case c of
    Arithmetic  a   ->  let res = fst $ encodeA f j a
                            inc = snd $ encodeA f j a
                        in encode' f (j+inc) cs (res:out)
    Push        s i -> encode' f j cs ((encodePU f s i):out)
    Pop         s i -> encode' f j cs ((encodePO f s i):out)
    Label       s   -> []
    Goto        s   -> []
    If          s   -> []
    Function    s i -> []
    Call        s i -> []
    Return          -> []

encodeA :: BaseName -> Addr -> Arithmetic -> (T.Text, Addr)
encodeA f j a = bimap T.unlines fromInteger (case a of
    Add -> ([preBinaryOp, "M=D+M"]               , 0)
    Sub -> ([preBinaryOp, "M=M-D"]               , 0)
    Neg -> ([preUnaryOp,  "M=-M"]                , 0)
    Eq  -> ([preBinaryOp, comparisonOp f j "JEQ"], 2)
    Gt  -> ([preBinaryOp, comparisonOp f j "JGT"], 2)
    Lt  -> ([preBinaryOp, comparisonOp f j "JLT"], 2)
    And -> ([preBinaryOp, "M=D&M"]               , 0)
    Or  -> ([preBinaryOp, "M=D|M"]               , 0)
    Not -> ([preUnaryOp,  "M=!M"]                , 0)
    )
    where
        preBinaryOp = T.unlines
            [ "@SP"
            , "AM=M-1"
            , "D=M"
            , "A=A-1"
            ]
        preUnaryOp = T.unlines
            [ "@SP"
            , "A=M-1"
            ]
        comparisonOp f j jump = T.unlines
            [ "D=M-D"
            , atVar f j
            , "D;" `T.append` jump
            , "@SP"
            , "A=M-1"
            , "M=0"
            , atVar f (j+1)
            , "0;JMP"
            , labelVar f j
            , "@SP"
            , "A=M-1"
            , "M=-1"
            , labelVar f (j+1)
            ]

encodePU :: BaseName -> Segment -> Addr -> T.Text
encodePU f s i = T.unlines $ (case s of
    Constant -> toD i
    Local    -> toDM "LCL" i
    Argument -> toDM "ARG" i
    This     -> toDM "THIS" i
    That     -> toDM "THAT" i
    Pointer  -> toDA 3 i
    Temp     -> toDA 5 i
    Static   ->
        [ atVar f i
        , "D=M"
        ]
    ) ++ pushD
    where
        toD i =
            [ atConst i
            , "D=A"
            ]
        toDM seg i = toD i ++
            [ atLabel seg
            , "A=D+M"
            , "D=M"
            ]
        toDA a i = toD i ++
            [ atConst a
            , "A=D+A"
            , "D=M"
            ]
        pushD =
            [ "@SP"
            , "A=M"
            , "M=D"
            , "@SP"
            , "M=M+1"
            ]

encodePO :: BaseName -> Segment -> Addr -> T.Text
encodePO f s i = T.unlines (case s of
    Constant -> popD
    Local    -> setRM "LCL" i   ++ popD ++ toR_D
    Argument -> setRM "ARG" i   ++ popD ++ toR_D
    This     -> setRM "THIS" i  ++ popD ++ toR_D
    That     -> setRM "THAT" i  ++ popD ++ toR_D
    Pointer  -> setRA 3 i       ++ popD ++ toR_D
    Temp     -> setRA 5 i       ++ popD ++ toR_D
    Static   -> popD ++
        [ atVar f i
        , "M=D"
        ]
    )
    where
        toD i =
            [ atConst i
            , "D=A"
            ]
        setRM seg i = toD i ++
            [ atLabel seg
            , "D=D+M"
            , "@R13"
            , "M=D"
            ]
        setRA a i = toD i ++
            [ atConst a
            , "D=D+A"
            , "@R13"
            , "M=D"
            ]
        popD =
            [ "@SP"
            , "AM=M-1"
            , "D=M"
            ]
        toR_D =
            [ "@R13"
            , "A=M"
            , "M=D"
            ]

atConst :: Addr -> T.Text
atConst a = T.pack $ '@' : showInt a ""

atVar :: BaseName -> Addr -> T.Text
atVar f j = T.pack $ '@' : f ++ '.' : showInt j ""

atLabel :: T.Text -> T.Text
atLabel l = '@' `T.cons` l

label :: T.Text -> T.Text
label l = T.concat ["(", l, ")"]

labelVar :: BaseName -> Addr -> T.Text
labelVar f j = T.pack $ '(' : f ++ '.' : showInt j ")"
